### ISSUE TITLE
CUDA ERROR running h2ogpt container on server

### ISSUE THREAD

**Issue by NicholasGoh** [h2oai/h2ogpt#771]
#### Issue description

h2ogpt docker container works locally but not on server

#### Steps to reproduce the issue

```
docker pull gcr.io/vorvan/h2oai/h2ogpt-runtime@sha256:2a74f35fd25069fbd8567512465c2b5d2c2b4211a67ed1d80aa62c5922bf0da9
docker tag gcr.io/vorvan/h2oai/h2ogpt-runtime@sha256:2a74f35fd25069fbd8567512465c2b5d2c2b4211a67ed1d80aa62c5922bf0da9 gcr.io/vorvan/h2oai/h2ogpt-runtime:0.1.0
```

Shared folder has all the weights pre-downloaded for running in offline server:

```
SHARED_FOLDER=/shared_folder
docker run \
    -e HOME=/workspace \
    -e HF_DATASETS_OFFLINE=1 \
    -e TRANSFORMERS_OFFLINE=1 \
    --gpus "device=0" \
    --runtime=nvidia \
    --shm-size=2g \
    -p 7860:7860 \
    --rm --init \
    -v $SHARED_FOLDER/.cache:/workspace/.cache \
    -v $SHARED_FOLDER/save:/workspace/save \
    gcr.io/vorvan/h2oai/h2ogpt-runtime:0.1.0 /workspace/generate.py \
        --user_path=vector_search \
        --base_model=TheBloke/Llama-2-7b-Chat-GPTQ \
        --model_path_llama=/workspace/.cache/huggingface/model_path_llama/llama-2-7b-chat.ggmlv3.q8_0.bin \
        --prompt_type=instruct \
        --score_model=None \
        --load_gptq=model \
        --use_safetensors=True \
        --langchain_mode=UserData \
        --user_path=vector_search \
        --num_async=10 \
        --gradio_offline_level=2 \
        --local_files_only=True
``` 

Any query into the UI leads to the errors listed in the next few sections

#### What's the expected result?

- Response generated after query

#### What's the actual result?

Snippet from full logs 

```
RuntimeError: CUDA error: no kernel image is available for execution on the device
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

#### Additional details

Compute architecture used is `sm_70`

```
nic@v100-server:~$ nvidia-smi --query-gpu=gpu_name,compute_cap,driver_version --format=csv
name, compute_cap, driver_version
Tesla V100-SXM2-32GB, 7.0, 530.30.02
```

Which is supported by the pytorch in the h2o container:

```
h2ogpt@802ea709438e:~# python -c "import torch; import sys; print('sys.version:', sys.version); print('torch.__version__:', torch.__version__); print('torch.cuda.is_available():', torch.cuda.is_available()); print('torch.backends.cudnn.enabled:', torch.backends.cudnn.enabled); device = torch.device('cuda'); print('torch.cuda.get_device_properties(device)', torch.cuda.get_device_properties(device)); print('torch.tensor([1.0, 2.0]).cuda():', torch.tensor([1.0, 2.0]).cuda()); print('torch.cuda.get_arch_list():', torch.cuda.get_arch_list())"
sys.version: 3.10.9 (main, Jan 11 2023, 15:21:40) [GCC 11.2.0]
torch.__version__: 2.0.1+cu117
torch.cuda.is_available(): True
torch.backends.cudnn.enabled: True
torch.cuda.get_device_properties(device) _CudaDeviceProperties(name='Tesla V100-SXM2-32GB', major=7, minor=0, total_memory=32501MB, multi_processor_count=80)
torch.tensor([1.0, 2.0]).cuda(): tensor([1., 2.], device='cuda:0')
torch.cuda.get_arch_list(): ['sm_37', 'sm_50', 'sm_60', 'sm_70', 'sm_75', 'sm_80', 'sm_86']

```

The same steps above run with no error locally on my RTX4060 despite no support for `sm_89`:

```
nic@local:~$ nvidia-smi --query-gpu=gpu_name,compute_cap,driver_version --format=csv                                                                                                                                        
name, compute_cap, driver_version
NVIDIA GeForce RTX 4060 Laptop GPU, 8.9, 525.125.06
```

```
h2ogpt@fd9f5c11eb4b:~# python -c "import torch; import sys; print('sys.version:', sys.version); print('torch.__version__:', torch.__version__); print('torch.cuda.is_available():', torch.cuda.is_available()); print('torch.backends.cudnn.enabled:', torch.backends.cudnn.enabled); device = torch.device('cuda'); print('torch.cuda.get_device_properties(device)', torch.cuda.get_device_properties(device)); print('torch.tensor([1.0, 2.0]).cuda():', torch.tensor([1.0, 2.0]).cuda()); print('torch.cuda.get_arch_list():', torch.cuda.get_arch_list())"
sys.version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0]
torch.__version__: 2.0.1+cu117
torch.cuda.is_available(): True
torch.backends.cudnn.enabled: True
torch.cuda.get_device_properties(device) _CudaDeviceProperties(name='NVIDIA GeForce RTX 4060 Laptop GPU', major=8, minor=9, total_memory=7917MB, multi_processor_count=24)
torch.tensor([1.0, 2.0]).cuda(): tensor([1., 2.], device='cuda:0')
torch.cuda.get_arch_list(): ['sm_37', 'sm_50', 'sm_60', 'sm_70', 'sm_75', 'sm_80', 'sm_86']
```

**Comment by pseudotensor** [comment]
It may be llama_cpp_python build issue.

But a question, can you use huggingface models instead of GGML?
```
--base_model=TheBloke/Llama-2-7b-Chat-GPTQ \
        --model_path_llama=/workspace/.cache/huggingface/model_path_llama/llama-2-7b-chat.ggmlv3.q8_0.bin \
```

Also, you have the --prompt_type=instruct, but for llama-2 it should be `--prompt_type=llama2`

**Comment by pseudotensor** [comment]
@achraf-mer @ChathurindaRanasinghe When I tried his command on latest image I get:
```
Starting get_model: TheBloke/Llama-2-7b-Chat-GPTQ 
/h2ogpt_conda/lib/python3.10/site-packages/transformers/models/auto/configuration_auto.py:995: FutureWarning: The `use_auth_token` argument is deprecated and will be removed in v5 of Transformers.
  warnings.warn(
Traceback (most recent call last):
  File "/workspace/generate.py", line 16, in <module>
    entrypoint_main()
  File "/workspace/generate.py", line 12, in entrypoint_main
    H2O_Fire(main)
  File "/workspace/src/utils.py", line 60, in H2O_Fire
    fire.Fire(component=component, command=args)
  File "/h2ogpt_conda/lib/python3.10/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/h2ogpt_conda/lib/python3.10/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/h2ogpt_conda/lib/python3.10/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/workspace/src/gen.py", line 1052, in main
    model0, tokenizer0, device = get_model(reward_type=False,
  File "/workspace/src/gen.py", line 1408, in get_model
    config, _, max_seq_len = get_config(base_model, **config_kwargs, raise_exception=False)
  File "/workspace/src/gen.py", line 1123, in get_config
    config = AutoConfig.from_pretrained(base_model, use_auth_token=use_auth_token,
  File "/h2ogpt_conda/lib/python3.10/site-packages/transformers/models/auto/configuration_auto.py", line 1007, in from_pretrained
    config_dict, unused_kwargs = PretrainedConfig.get_config_dict(pretrained_model_name_or_path, **kwargs)
  File "/h2ogpt_conda/lib/python3.10/site-packages/transformers/configuration_utils.py", line 620, in get_config_dict
    config_dict, kwargs = cls._get_config_dict(pretrained_model_name_or_path, **kwargs)
  File "/h2ogpt_conda/lib/python3.10/site-packages/transformers/configuration_utils.py", line 675, in _get_config_dict
    resolved_config_file = cached_file(
  File "/h2ogpt_conda/lib/python3.10/site-packages/transformers/utils/hub.py", line 428, in cached_file
    resolved_file = hf_hub_download(
  File "/h2ogpt_conda/lib/python3.10/site-packages/huggingface_hub/utils/_validators.py", line 118, in _inner_fn
    return fn(*args, **kwargs)
  File "/h2ogpt_conda/lib/python3.10/site-packages/huggingface_hub/file_download.py", line 1159, in hf_hub_download
    os.makedirs(storage_folder, exist_ok=True)
  File "/h2ogpt_conda/lib/python3.10/os.py", line 215, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/h2ogpt_conda/lib/python3.10/os.py", line 215, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/h2ogpt_conda/lib/python3.10/os.py", line 225, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/workspace/.cache/huggingface'
```

Even though my ~/.cache/huggingface has fine permissions.

Is that expected?

**Comment by NicholasGoh** [comment]
I managed to fix this issue by upgrading to `auto-gptq==0.4.2` from `0.3.0`, please close this issue at your convenience

```
pip install --force-reinstall --no-deps auto-gptq==0.4.2
```

regarding permission issues, running docker as root `-u root`  or giving the directory permissions for user `1000` i.e user `h2ogpt` should work as well

as for the prompt type, i have had no issues with `llama2` or `instruct`

also, specifying the GGML file was not needed, my apologies

**Comment by pseudotensor** [comment]
Ok, auto-gptq==0.4.2 is latest we use.

Ok, instruct will be worse than llama2, but llama2 model is kinda robust to bad prompt format.
